dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/IttProvider.cshtml.cs

```diff
@@ -0,0 +1,82 @@
+using System.ComponentModel.DataAnnotations;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using TeacherIdentity.AuthServer.Services.DqtApi;
+
+namespace TeacherIdentity.AuthServer.Pages.SignIn.Trn;
+
+[BindProperties]
+public class IttProvider : PageModel
+{
+    private IIdentityLinkGenerator _linkGenerator;
+    private IDqtApiClient _dqtApiClient;
+
+    public IttProvider(
+        IIdentityLinkGenerator linkGenerator,
+        IDqtApiClient dqtApiClient)
+    {
+        _linkGenerator = linkGenerator;
+        _dqtApiClient = dqtApiClient;
+    }
+
+    public string[]? IttProviderNames;
+
+    // Properties are set in the order that they are declared. Because the value of AwardedByTrainer
+    // is used in the conditional RequiredIfTrue attribute, it should be set first.
+    [Display(Name = "Did a university, SCITT or school award your QTS?")]
+    [Required(ErrorMessage = "Tell us how you were awarded qualified teacher status (QTS)")]
+    public bool? HasIttProvider { get; set; }
+
+    [Display(Name = "Where did you get your QTS?", Description = "Your university, SCITT, school or other training provider")]
+    [RequiredIfTrue(nameof(HasIttProvider), ErrorMessage = "Enter your university, SCITT, school or other training provider")]
+    public string? IttProviderName { get; set; }
+
+    public void OnGet()
+    {
+    }
+
+    public IActionResult OnPost()
+    {
+        if (!ModelState.IsValid)
+        {
+            return this.PageWithErrors();
+        }
+
+        HttpContext.GetAuthenticationState().HaveIttProvider = HasIttProvider;
+
+        if (HasIttProvider == true)
+        {
+            HttpContext.GetAuthenticationState().IttProviderName = IttProviderName;
+        }
+
+        return Redirect(_linkGenerator.TrnCheckAnswers());
+    }
+
+    public override async Task OnPageHandlerExecutionAsync(PageHandlerExecutingContext context, PageHandlerExecutionDelegate next)
+    {
+        var authenticationState = context.HttpContext.GetAuthenticationState();
+
+        // We expect to have a verified email and official names at this point but we shouldn't have completed the TRN lookup
+        if (string.IsNullOrEmpty(authenticationState.EmailAddress) ||
+            !authenticationState.EmailAddressVerified ||
+            !authenticationState.HasOfficialName() ||
+            authenticationState.HaveCompletedTrnLookup)
+        {
+            context.Result = new RedirectResult(authenticationState.GetNextHopUrl(_linkGenerator));
+            return;
+        }
+
+        if (HttpContext.Session.Get<string[]>("IttProviderNames") == null)
```

Comment:
This prevents us from calling the API again when e.g. the form is submitted

Reply:
Might be worth stashing this in `IMemoryCache`; iirc it's not the fastest endpoint and it's relatively static data.